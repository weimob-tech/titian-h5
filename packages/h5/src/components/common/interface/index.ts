export type OmitType<T, P> = Pick<T, { [key in keyof T]: T[key] extends P ? never : key }[keyof T]>;

export type Key = string | number | symbol;

export type IAnyObject = { [key: Key]: unknown };

export type IAnyMap = Map<Key, unknown>;

export type DistanceUnit = 'px' | 'em' | 'rem' | 'vw' | 'vh' | '%';

export type IPosition = {
  bottom?: `${number}${DistanceUnit}`;
  top?: `${number}${DistanceUnit}`;
  left?: `${number}${DistanceUnit}`;
  right?: `${number}${DistanceUnit}`;
};
