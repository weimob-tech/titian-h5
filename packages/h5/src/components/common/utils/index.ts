import { getTitianMode } from '../../../global/global';
import { IAnyObject, OmitType } from '../interface/index';

export * from './number';

export const objectToString = Object.prototype.toString;

export const toTypeString = (val: unknown): string => objectToString.call(val);

export type Key = string | number | symbol;

export const isString = (obj: unknown): obj is string => toTypeString(obj) === '[object String]';

export const isPlainObject = (obj: unknown): obj is { [key: Key]: unknown } => toTypeString(obj) === '[object Object]';

export const isPlainArray = (obj: unknown): obj is unknown[] => toTypeString(obj) === '[object Array]';

export const isPlainRegExp = (obj: unknown): obj is RegExp => toTypeString(obj) === '[object RegExp]';

export const isBoolean = (obj: unknown): obj is boolean => toTypeString(obj) === '[object Boolean]';

export const isFunction = (obj: unknown): obj is (...agrn: unknown[]) => unknown => typeof obj === 'function';

export const isArrayEqual = (arr1: unknown[], arr2: unknown[]): boolean => {
  if (arr1.length !== arr2.length) {
    return false;
  }

  return arr1.every((item, index) => item === arr2[index]);
};

export function debounce(fn: (...agrn: unknown[]) => unknown, ms = 300) {
  let timeoutId: ReturnType<typeof setTimeout>;
  return function debounceFn(this: unknown, ...args: unknown[]) {
    clearTimeout(timeoutId);
    timeoutId = setTimeout(() => fn.apply(this, args), ms);
  };
}

export function throttle(fn: (...agrn: unknown[]) => unknown, ms = 300) {
  let inThrottle: boolean;
  let lastFn: ReturnType<typeof setTimeout>;
  let lastTime: number;
  return function r(this: unknown, ...args: unknown[]) {
    if (!inThrottle) {
      fn.apply(this, args);
      lastTime = Date.now();
      inThrottle = true;
    } else {
      clearTimeout(lastFn);
      lastFn = setTimeout(() => {
        if (Date.now() - lastTime >= ms) {
          fn.apply(this, args);
          lastTime = Date.now();
        }
      }, Math.max(ms - (Date.now() - lastTime), 0));
    }
  };
}

export const randomString = () => Math.random().toString(36).slice(2);

export function filterInvalidData<T extends IAnyObject>(obj: T): OmitType<T, null | undefined> {
  const params = { ...obj };
  Object.keys(obj).forEach(key => {
    const value = obj[key];
    if (typeof value === 'undefined' || value === null) {
      delete params[key];
    }
  });
  return params;
}

export const DISABLED = 'disabled';

export function isDisabled(params: unknown): boolean {
  if (typeof params !== 'object' || !params) {
    return false;
  }
  return (params as IAnyObject)[DISABLED] as boolean;
}

export function range(num: number, min: number, max: number) {
  return Math.min(Math.max(num, min), max);
}

export function padZero(val: string | number, number = 2) {
  return `${val}`.padStart(number, '0');
}

export function isValidDate(d: unknown): d is Date {
  return d !== undefined && d !== null && !Number.isNaN(new Date(d as Date).getTime());
}

export function getDate(): Date;
export function getDate(date: string | number | Date): Date;
export function getDate(year: number, month: number): Date;
export function getDate(year: number, month: number, day: number): Date;
export function getDate(date?: string | number | Date, month?: number, day?: number) {
  let cur: Date;
  if (typeof date === 'undefined') {
    return new Date();
  }
  if (typeof date === 'number' && typeof month === 'number' && typeof day === 'number') {
    cur = new Date(date, month, day);
  } else if (typeof date === 'number' && typeof month === 'number') {
    cur = new Date(date, month);
  } else {
    cur = new Date(date);
  }
  if (!cur) {
    throw new Error('时间格式错误');
  }
  if (cur.toString() === 'Invalid Date') {
    throw new Error('时间格式错误');
  }
  return cur;
}

export function isUndefined(val: unknown) {
  return toTypeString(val) === '[object Undefined]';
}

// eslint-disable-next-line @titian-design/no-more-args
export function easeInOutCubic(t: number, b: number, c: number, d: number) {
  const cc = c - b;
  t /= d / 2;
  if (t < 1) {
    return (cc / 2) * t * t * t + b;
  }
  return (cc / 2) * ((t -= 2) * t * t + 2) + b;
}

// eslint-disable-next-line @titian-design/no-more-args
export function addEventListener<K extends Element>(
  target: K,
  eventType: string,
  cb: (arg: any) => any,
  options?: boolean | AddEventListenerOptions,
): () => void {
  if (target.addEventListener) {
    target.addEventListener(eventType, cb, options);
    document.addEventListener(eventType, cb, options);
  }

  return () => {
    if (target.removeEventListener) {
      target.removeEventListener(eventType, cb, options);
    }
  };
}

export const attrStyleToString = (attr: unknown) => {
  if (!attr) {
    return '';
  }
  if (typeof attr === 'string') {
    return attr;
  }
  if (Array.isArray(attr)) {
    return attr.join(';');
  }
  if (Object.keys(attr as { [key: string]: unknown }).length > 0) {
    return Object.keys(attr as { [key: string]: unknown })
      .reduce((array, key) => {
        array.push(`${key}:${(attr as { [key: string]: unknown })[key]}`);
        return array;
      }, [] as string[])
      .join(';');
  }

  return '';
};

export const stringToAttrStyle = (attrStr: unknown): { [key: string]: string } => {
  if (!attrStr) {
    return {};
  }
  if (isPlainObject(attrStr)) {
    return attrStr as { [key: string]: string };
  }
  if (!isString(attrStr)) {
    return {};
  }
  try {
    attrStr = JSON.parse(attrStr);
    // eslint-disable-next-line no-empty
  } catch (e) {}
  if (isPlainObject(attrStr)) {
    return attrStr as { [key: string]: string };
  }
  if (!isString(attrStr)) {
    return {};
  }
  return attrStr
    .trim()
    .split(';')
    .filter(Boolean)
    .reduce((pre, cur) => {
      const [key, value] = cur.split(':');
      const nkey = key.trim();
      pre[nkey] = value;
      return pre;
    }, {} as { [key: string]: string });
};

export const getResponsivePixel = (px: number) => {
  const mode = getTitianMode();
  if (mode === 'pc') {
    return px / 2;
  }
  return (document.documentElement.clientWidth / 750) * px;
};

export function getStyle<T extends HTMLElement>(element: T, attrName: keyof CSSStyleDeclaration) {
  let ret = element.style[attrName];
  if (ret) {
    return ret;
  }

  const getComputedStyle = document?.defaultView?.getComputedStyle;

  if (getComputedStyle) {
    ret = getComputedStyle(element, null)[attrName];
  }
  return ret;
}

export function setStyle(
  dom: HTMLElement,
  style: Partial<{ [key in keyof CSSStyleDeclaration]: CSSStyleDeclaration[key] | null }>,
) {
  (Object.keys(style) as (keyof CSSStyleDeclaration)[]).forEach(key => {
    if (style[key] === null) {
      dom.style.removeProperty(key as string);
    } else {
      dom.style.setProperty(key as string, style[key] as string);
    }
  });
}
export function hasClass(target: Element, className: string) {
  return target.className.indexOf(className) > -1;
}
export function addClass(target: Element, className: string) {
  return target.classList.add(className);
}
export function removeClass(target: Element, className: string) {
  return target.classList.remove(className);
}
export function toggleClass(target: Element, className: string) {
  return target.classList.toggle(className);
}
export function isCustomEvent(event: CustomEvent | Event): event is CustomEvent {
  return !event.isTrusted;
}

//  end 向后 无法查找 则  start end 之间
export function nextAvailable<T>(params: {
  start: number;
  end: number;
  list: T[];
  fn?: (list: T[], num: number) => boolean;
}) {
  const { start, end, list, fn } = params;
  let num = end;
  while (num < list.length) {
    num += 1;
    if (typeof list[num] !== undefined) {
      if (typeof fn === 'function' && fn(list, num)) {
        return num;
      }
    }
  }
  num = end;
  while (num > start) {
    num -= 1;
    if (typeof list[num] !== undefined) {
      if (typeof fn === 'function' && fn(list, num)) {
        return num;
      }
    }
  }
  return -1;
}

//  end 向前 无法查找 则 start end 之间
export function prevAvailable<T>(params: {
  start: number;
  end: number;
  list: T[];
  fn?: (list: T[], num: number) => boolean;
}) {
  const { start, list, end, fn } = params;
  let num = end;
  // 节点向前
  while (num > 0) {
    num -= 1;
    if (typeof list[num] !== undefined) {
      if (typeof fn === 'function' && fn(list, num)) {
        return num;
      }
    }
  }
  num = end;
  while (num < start) {
    num += 1;
    if (typeof list[num] !== undefined) {
      if (typeof fn === 'function' && fn(list, num)) {
        return num;
      }
    }
  }
  return -1;
}

export function addShadowRootStyle() {
  if (!this.extCss || !this.host || !this.host.shadowRoot) return;
  const style = this.host.shadowRoot.querySelector('style') || document.createElement('style');
  style.innerHTML = this.extCss;
  this.host.shadowRoot.appendChild(style);
}

export function pxToVW(str) {
  const mode = getTitianMode();

  if (mode === 'pc') return str;

  const res = (Number(str.replace('px', '')) * 100) / 750;
  function toFixedFn(value, n) {
    return Math.round(value * 10 ** n) / 10 ** n;
  }
  return `${toFixedFn(res, 5)}vw`;
}

export function validateEvent(event, shadowRoot) {
  const root = event.target.getRootNode();
  if (root === shadowRoot) {
    return true;
  }
  const slot = event.path.find(element => element.tagName === 'SLOT');
  const [node] = slot.assignedNodes();
  if (node && node.hasAttribute('slot')) {
    const stop = node.getAttribute('stopPropagation') || node.getAttribute('stop-propagation');
    if (typeof stop === 'boolean' && stop) {
      return false;
    }
    if (typeof stop === 'string' && (stop === 'true' || stop === '')) {
      return false;
    }
  }
  return true;
}

export function checkSlotIsUsed(host) {
  const slot = host.shadowRoot?.querySelector('slot')?.assignedNodes({ flatten: true });
  if (!slot || slot.length === 0) return false;
  if (slot && slot.length === 1) {
    if (slot[0].nodeName === '#text' && slot[0].textContent === '') {
      return false;
    }
  }
  return true;
}
