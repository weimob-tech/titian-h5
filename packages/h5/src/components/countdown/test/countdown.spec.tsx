import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { TiCountdown } from '../index';
import { sleep } from '../../common/test';

const time = 11118888666;
describe('渲染 ti-countdown', () => {
  it('基本渲染', async () => {
    const page = await newSpecPage({
      components: [TiCountdown],
      template: () => <ti-countdown autoplay time={time} variant="mixture" format="DD天HH时mm分ss SSS" size="big" />,
    });
    page.root.setAttribute('time', '1118888666');

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.body.innerHTML = '';
  });

  it('格式化时间', async () => {
    const page = await newSpecPage({
      components: [TiCountdown],
      template: () => <ti-countdown time={time} format="SSS" />,
    });
    page.root.setAttribute('format', 'HH:mm:ss SSS');
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.body.innerHTML = '';
  });

  it('风格', async () => {
    const page = await newSpecPage({
      components: [TiCountdown],
      template: () => <ti-countdown autoplay time={time} variant="pure" format="DD天HH:mm:ss SSS" />,
    });

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.body.innerHTML = '';
  });

  it('暂停', async () => {
    const time = 2000;
    const page = await newSpecPage({
      components: [TiCountdown],
      template: () => <ti-countdown time={time} variant="pure" format="DD天HH:mm:ss SSS"></ti-countdown>,
    });
    await page.root.start();
    await sleep(1000);
    await page.root.pause();
    await sleep(200);
    await page.root.start();
    await sleep(2000);
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.body.innerHTML = '';
  });

  it('尺寸', async () => {
    const time = 1200;
    const page = await newSpecPage({
      components: [TiCountdown],
      template: () => <ti-countdown autoplay useSlot time={time} variant="block" size="small" />,
    });
    page.root.setAttribute('use-slot', 'false');
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.body.innerHTML = '';
  });

  it('自定义内容', async () => {
    const timeGroup = { day: '', hour: '', minute: '', second: '', millisecond: '' };
    const page = await newSpecPage({
      components: [TiCountdown],
      template: () => (
        <ti-countdown time={time}>
          <div class="time">{timeGroup.day}</div>
          <div class="tag">天</div>
          <div class="time">{timeGroup.hour}</div>
          <div class="tag">:</div>
          <div class="time">{timeGroup.minute}</div>
          <div class="tag">:</div>
          <div class="time">{timeGroup.second}</div>
          <div class="tag">:</div>
          <div class="time">{timeGroup.millisecond}</div>
        </ti-countdown>
      ),
    });
    page.root.setAttribute('use-slot', 'true');
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
    page.body.innerHTML = '';
  });
});
