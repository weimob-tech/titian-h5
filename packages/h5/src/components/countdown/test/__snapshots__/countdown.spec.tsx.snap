// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`渲染 ti-countdown 基本渲染 1`] = `
<ti-countdown time="1118888666">
  <mock:shadow-root>
    <div class="titian-countdown undefined">
      <div class="titian-countdown-box titian-countdown-box-big">
        <div class="titian-countdown-mixture">
          12
        </div>
        <div class="titian-countdown-day titian-countdown-day-mixture">
          天
        </div>
        <div class="titian-countdown-block">
          22
        </div>
        <div class="titian-countdown-dot">
          <ti-icon name="colon"></ti-icon>
        </div>
        <div class="titian-countdown-block">
          48
        </div>
        <div class="titian-countdown-dot">
          <ti-icon name="colon"></ti-icon>
        </div>
        <div class="titian-countdown-block">
          08
        </div>
        <div class="titian-countdown-dot">
          <ti-icon name="colon"></ti-icon>
        </div>
        <div class="titian-countdown-block titian-countdown-block-millisecond">
          666
        </div>
      </div>
    </div>
  </mock:shadow-root>
</ti-countdown>
`;

exports[`渲染 ti-countdown 尺寸 1`] = `
<ti-countdown use-slot="false">
  <mock:shadow-root>
    <div class="titian-countdown undefined">
      <div class="titian-countdown-box titian-countdown-box-small">
        <div class="titian-countdown-block"></div>
        <div class="titian-countdown-day titian-countdown-day-block">
          天
        </div>
        <div class="titian-countdown-block">
          00
        </div>
        <div class="titian-countdown-dot">
          <ti-icon name="colon"></ti-icon>
        </div>
        <div class="titian-countdown-block">
          00
        </div>
        <div class="titian-countdown-dot">
          <ti-icon name="colon"></ti-icon>
        </div>
        <div class="titian-countdown-block">
          01
        </div>
        <div class="titian-countdown-dot">
          <ti-icon name="colon"></ti-icon>
        </div>
        <div class="titian-countdown-block titian-countdown-block-millisecond"></div>
      </div>
    </div>
  </mock:shadow-root>
</ti-countdown>
`;

exports[`渲染 ti-countdown 暂停 1`] = `
<ti-countdown>
  <mock:shadow-root>
    <div class="titian-countdown undefined">
      00天00:00:00 000
    </div>
  </mock:shadow-root>
</ti-countdown>
`;

exports[`渲染 ti-countdown 格式化时间 1`] = `
<ti-countdown format="HH:mm:ss SSS">
  <mock:shadow-root>
    <div class="titian-countdown undefined">
      16:34:48 666
    </div>
  </mock:shadow-root>
</ti-countdown>
`;

exports[`渲染 ti-countdown 自定义内容 1`] = `
<ti-countdown use-slot="true">
  <mock:shadow-root>
    <div class="titian-countdown undefined">
      <slot></slot>
    </div>
  </mock:shadow-root>
  <div class="time"></div>
  <div class="tag">
    天
  </div>
  <div class="time"></div>
  <div class="tag">
    :
  </div>
  <div class="time"></div>
  <div class="tag">
    :
  </div>
  <div class="time"></div>
  <div class="tag">
    :
  </div>
  <div class="time"></div>
</ti-countdown>
`;

exports[`渲染 ti-countdown 风格 1`] = `
<ti-countdown>
  <mock:shadow-root>
    <div class="titian-countdown undefined">
      128天16:34:48 666
    </div>
  </mock:shadow-root>
</ti-countdown>
`;
